### PR TITLE
add a design doc for switchless

### PR DESCRIPTION
**Pull request by jxyang**

### CODE REVIEW

File: docs/DesignDocs/SwitchlessCalls.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+## Motivation
+
+Context-switchless Calls are designed to reduce the cost of context switching between hosts and enclaves.
+In an enclave application, the host makes **ECALL**s into functions exposed by the enclaves it created. Likewise,
+the enclaves may make **OCALL**s into functions exposed by the host that created them. In either case, the
+execution has to be transitioned from an untrusted environment to a trusted environment, or vice versa. Since the
+transition is costly due to heavy security checks, it might be more performance advantageous to make the calls
+**context-switchless**: the caller delegates the function call to a worker thread in the other environment, which
+does the real job of calling the function and post the result to the caller. Both the calling thread and the
```
Comment (dthaler) [inline docs/DesignDocs/SwitchlessCalls.md]: grammar: "post" -> "posting"

File: docs/DesignDocs/SwitchlessCalls.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+worker thread never leave their respective execution contexts during the perceived function call.
+
+## Possible Usages
+
+In general, the good candidates for switchless calls are functions that are:
+1) short, thus the transition takes relatively high percentage of the overall execution time of the call; and
+2) called frequently, so the savings in transition time add up.
```
Comment (dthaler) [inline docs/DesignDocs/SwitchlessCalls.md]: grammar: "add" -> "adds"

File: docs/DesignDocs/SwitchlessCalls.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+execution has to be transitioned from an untrusted environment to a trusted environment, or vice versa. Since the
+transition is costly due to heavy security checks, it might be more performance advantageous to make the calls
+**context-switchless**: the caller delegates the function call to a worker thread in the other environment, which
+does the real job of calling the function and post the result to the caller. Both the calling thread and the
+worker thread never leave their respective execution contexts during the perceived function call.
+
+## Possible Usages
+
+In general, the good candidates for switchless calls are functions that are:
+1) short, thus the transition takes relatively high percentage of the overall execution time of the call; and
```
Comment (dthaler) [inline docs/DesignDocs/SwitchlessCalls.md]: grammar: "high" -> "a high"

File: docs/DesignDocs/SwitchlessCalls.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+2) called frequently, so the savings in transition time add up.
+
+If your enclave application has such functions and you are concerned with performance, consider making those
+functions context-switchless.
+
+## User Experience
+
+Firstly, The user has to identify which functions are good candidates of switchless calls. The identified ones
```
Comment (dthaler) [inline docs/DesignDocs/SwitchlessCalls.md]: capitalization: "Firstly, The" -> "Firstly, the"

File: docs/DesignDocs/SwitchlessCalls.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+need to be marked in the EDL file with a keyword `transition_using_threads`. For example, to mark function
+`host_increment_switchless` a target of switchless calls, it has to be declared as:
+
+```c
+void host_increment_switchless([in, out] int* m) transition_using_threads;
+```
+
+Secondly, while creating an enclave, the user has to explicitly configure it to enable switchless capability.
+An important setting in the configuration is how many worker threads are to be created for servicing the
+context-switchless calls. More worker threads typically means more competition for the CPU cores and more thread context switches, hurting the performance. On the other hand, less worker threads means means simultaneously issued switchless calls are less likely to be serviced quickly, if they got serviced at all.
```
Comment (dthaler) [inline docs/DesignDocs/SwitchlessCalls.md]: grammar: "less worker threads" -> "fewer worker threads"
(Also would be good to break this line into multiple lines in the source to make diffs easier to review)

File: docs/DesignDocs/SwitchlessCalls.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+We will give a guideline for users to search for the "sweet spot" of this setting.
+
+The users are encouraged to measure the performance delta between enclave applications with
+or without switchless calls, and decide on whether switchless call should be turned on for some functions,
```
Comment (dthaler) [inline docs/DesignDocs/SwitchlessCalls.md]: grammar: "call" -> "calls" or "calling", whichever you prefer

File: docs/DesignDocs/SwitchlessCalls.md (hunk 1 above)
Comment (jxyang) [reply docs/DesignDocs/SwitchlessCalls.md]: Thank you for the proofreading. @dthaler . All errors are fixed.

File: docs/DesignDocs/SwitchlessCalls.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
+Context-switchless Calls
+================
+
+## Motivation
+
+Context-switchless Calls are designed to reduce the cost of context switching between hosts and enclaves.
```
[the remark this answers is not in the document]
Comment (jxyang) [reply docs/DesignDocs/SwitchlessCalls.md]: The best practice, AFAIK, is to hypen-connect the first 2 words in a 3-word phrase, where the first two together is the adjective of the 3rd noun word.

File: docs/DesignDocs/SwitchlessCalls.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+In an enclave application, the host makes **ECALL**s into functions exposed by the enclaves it created. Likewise,
+the enclaves may make **OCALL**s into functions exposed by the host that created them. In either case, the
+execution has to be transitioned from an untrusted environment to a trusted environment, or vice versa. Since the
+transition is costly due to heavy security checks, it might be more performance advantageous to make the calls
```
[the remark this answers is not in the document]
Comment (jxyang) [reply docs/DesignDocs/SwitchlessCalls.md]: Added some explanations. Also added a `security considerations` section below.

**Pull request merged**